### ISSUE TITLE
Attach to stopped container with stream=1 never closes connection

### ISSUE THREAD

**Issue by msmith-techempower** [moby/moby#41922]
# What happens?

When attempting to attach to a **stopped** container via the [HTTP API](https://docs.docker.com/engine/api/v1.40/#operation/ContainerAttach) with `stream=1&stdout=1`, `stream=1&stderr=1`, or `stream=1&stdout=1&stderr=1` the connection is never closed and clients may wait indefinitely.

# What did I expect?

The connection to be closed similarly to attempting to attach to a stopped container *without* `stream=1`

# How can this be reproduced?

Use `curl` to attempt to attach to a stopped container (the container id I'm using is `cdefe2318d6d`; replace with one of yours that is valid and stopped):
```
curl -vv --unix-socket /var/run/docker.sock -X POST "http://localhost/v1.41/containers/cdefe2318d6d/attach?stream=1&stdout=1"
```
This will produce the following verbose output:
```
*   Trying /var/run/docker.sock:0...
* Connected to localhost (/var/run/docker.sock) port 80 (#0)
> POST /v1.41/containers/cdefe2318d6d/attach?stream=1&stdout=1 HTTP/1.1
> Host: localhost
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/vnd.docker.raw-stream
* no chunk, no close, no size. Assume close to signal end
<
```

Notice: the connection is not close. We can do the same thing without `stream=1` or with `logs=1` instead:
```
curl -vv --unix-socket /var/run/docker.sock -X POST "http://localhost/v1.41/containers/cdefe2318d6d/attach?stdout=1"
```
which produces the following:
```
*   Trying /var/run/docker.sock:0...
* Connected to localhost (/var/run/docker.sock) port 80 (#0)
> POST /v1.41/containers/cdefe2318d6d/attach?logs=1&stdout=1 HTTP/1.1
> Host: localhost
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/vnd.docker.raw-stream
* no chunk, no close, no size. Assume close to signal end
<
* Closing connection 0
```

The important bit is that `*Closing connection 0` after the `Assume close to signal end`.

# How would this happen in practice and why should this be changed?

In my case, it happens because of a race condition between a container that is started by software I am working on, that container exiting naturally, and the software attempting to attach to stream logging. Adding a check to see whether the container is running does not help because of the same race condition - it may be running after I start/check it, but by the time the attach call is made, the container may have stopped.

# Other examples

If you connect to a *running* container with `stream=1&stdout=1` it will correctly *not* close the connection; however, if the container exits, the connection is closed. This *seems* intended and correct, and probably should be the intent in the case where the container is already stopped when the attach request is made.

# `docker version`
```
Client: Docker Engine - Community
 Cloud integration: 1.0.7
 Version:           20.10.2
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        2291f61
 Built:             Mon Dec 28 16:17:34 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.2
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8891c58
  Built:            Mon Dec 28 16:15:28 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

# `docker info`
```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)
  scan: Docker Scan (Docker Inc., v0.5.0)

Server:
 Containers: 78
  Running: 2
  Paused: 0
  Stopped: 76
 Images: 47
 Server Version: 20.10.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.19.104-microsoft-standard
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 32
 Total Memory: 31.32GiB
 Name: docker-desktop
 ID: UJNW:BHJS:WE67:Y3IA:DG5J:H6EF:32UR:767O:X6QN:CCNL:PFIE:H67P
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 95
  Goroutines: 120
  System Time: 2021-01-22T19:20:27.9225667Z
  EventsListeners: 3
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio weight support
WARNING: No blkio weight_device support
WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

# TL;DR

I believe it is correct for the connection to be closed when `stream=1&stdout=1`, `stream=1&stderr=1`, or `stream=1&stdout=1&stderr=1` *and* the container is stopped.

**Comment by thaJeztah** [comment]
ISTR this may be by design, to allow creating a container, attach to it before it's started, to prevent missing initial output of the container.

@cpuguy83 perhaps you can confirm if I'm right?

**Comment by msmith-techempower** [comment]
> ISTR this may be by design, to allow creating a container, attach to it before it's started, to prevent missing initial output of the container.

Wouldn't the `logs=1` flag cover that situation? 

For my case, I start the container, then attempt to attach with `stream=1&logs=1&stdout=1&stderr=1` to get all the logs *and more* but only if the container is still running.

Also, (and I will retest this on my end) I did not think that I was able to attach to a not-yet-started container. Again, I will retest, and if that is the intent for my use case, then I will update and close.

**Comment by msmith-techempower** [comment]
Well, that worked. Thanks for the help!
